arch/x86/core/x86_mmu.c

```diff
@@ -46,76 +47,123 @@ int _arch_buffer_validate(void *addr, size_t size, int write)
-		 * would be 1023. So each pde will be using all the
-		 * page table entires except for the last pde.
-		 * For the last pde, pte is calculated using the last
-		 * memory address of the buffer.
-		 */
-		if (pde != end_pde_num) {
-			ending_pte_num = 1023;
+		if (pdpte != end_pdpte_num) {
+			end_pde_num = 0;
 		} else {
-			ending_pte_num = MMU_PAGE_NUM((char *)addr + size - 1);
+			end_pde_num = MMU_PDE_NUM((char *)addr + size - 1);
 		}
 
-		/* For all the pde's appart from the starting pde, will have
-		 * the start pte number as zero.
+		struct x86_mmu_page_directory *pd_address =
+			X86_MMU_GET_PD_ADDR_INDEX(pdpte);
+#endif
+		/* Iterate for all the pde's the buffer might take up.
+		 * (depends on the size of the buffer and start address
+		 * of the buff)
 		 */
-		if (pde != start_pde_num) {
-			starting_pte_num = 0;
-		}
-
-		pte_value.value = 0xFFFFFFFF;
-
-		/* Bitwise AND all the pte values. */
-		for (pte = starting_pte_num; pte <= ending_pte_num; pte++) {
-			pte_value.value &= pte_address->entry[pte].value;
-		}
+		for (pde = start_pde_num; pde <= end_pde_num; pde++) {
+#ifdef CONFIG_X86_PAE_MODE
+			union x86_mmu_pae_pde pde_value =
+				pd_address->entry[pde];
+#else
+			union x86_mmu_pde_pt pde_value =
+				X86_MMU_PD->entry[pde].pt;
+#endif
 
-		if (!pte_value.p || !pte_value.us || (write && !pte_value.rw)) {
-			return -EPERM;
+			if (!pde_value.p ||
+			    !pde_value.us ||
+			    (write && !pde_value.rw)) {
+				return -EPERM;
+			}
+
+			pte_address = (struct x86_mmu_page_table *)
+				(pde_value.page_table << MMU_PAGE_SHIFT);
+
+			/* loop over all the possible page tables for the
+			 * required size. If the pde is not the last one
+			 * then the last pte would be 1023. So each pde
+			 * will be using all the page table entires except
+			 * for the last pde. For the last pde, pte is
+			 * calculated using the last memory address
+			 * of the buffer.
+			 */
+			if (pde != end_pde_num) {
+				ending_pte_num = 1023;
+			} else {
+				ending_pte_num =
+					MMU_PAGE_NUM((char *)addr + size - 1);
+			}
+
+			/* For all the pde's appart from the starting pde,
+			 * will have the start pte number as zero.
+			 */
+			if (pde != start_pde_num) {
+				starting_pte_num = 0;
+			}
+
+			pte_value.value = 0xFFFFFFFF;
+
+			/* Bitwise AND all the pte values. */
+			for (pte = starting_pte_num;
+			     pte <= ending_pte_num;
+			     pte++) {
+				pte_value.value &=
+					pte_address->entry[pte].value;
+			}
+
+			if (!pte_value.p ||
+			    !pte_value.us ||
+			    (write && !pte_value.rw)) {
+				return -EPERM;
+			}
 		}
+#ifdef CONFIG_X86_PAE_MODE
 	}
+#endif
 
 	return 0;
 }
 
-
 static inline void tlb_flush_page(void *addr)
 {
 	/* Invalidate TLB entries corresponding to the page containing the
 	 * specified address
 	 */
 	char *page = (char *)addr;
+
 	__asm__ ("invlpg %0" :: "m" (*page));
 }
 
 
 void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
```

Comment:
seems like the "flags" and "mask" parameters need to be u64_t if PAE is enabled

Reply:
We don't need to read the XD bit in the PAE tables. So all the other flags are still within 32 bits. Hence I retained u32_t.
I want to keep all the implementation above this API agnostic of the width of the entries.
Makes the upper layers easier to use. I have added explicit typecasts to make the bit wise operations more meaningful.

Reply:
> We don't need to read the XD bit in the PAE tables

Not following you here.

1) How are we going to design a test case to verify that the page tables are setup correctly without being able to test at runtime with _x86_mmu_get_flags() if the XD bit is set?
2) How are we going to be able to support dynamically loaded application code if we are unable to set the XD bit on that application's text memory area at runtime?

> Makes the upper layers easier to use

This is not a public API exposed to applications, its used internally by the arch code. We have to have the capability to test/set the XD bit.